mempool/src/feerate.rs

```diff
@@ -0,0 +1,72 @@
+use std::num::NonZeroU128;
+
+use common::primitives::amount::Amount;
+
+lazy_static::lazy_static! {
+    pub(crate) static ref INCREMENTAL_RELAY_FEE_RATE: FeeRate = FeeRate::new(Amount::from_atoms(1000));
+    pub(crate) static ref INCREMENTAL_RELAY_THRESHOLD: FeeRate = FeeRate::new(Amount::from_atoms(500));
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct FeeRate {
+    atoms_per_kb: u128,
+}
+
+impl FeeRate {
+    pub(crate) fn new(tokens_per_kb: Amount) -> Self {
+        Self {
+            atoms_per_kb: tokens_per_kb.into_atoms(),
+        }
+    }
+
+    pub(crate) fn of_tx(fee: Amount, tx_size: usize) -> Self {
+        Self {
+            atoms_per_kb: Self::div_up(1000 * fee.into_atoms(), tx_size),
+        }
+    }
+
+    pub(crate) fn compute_fee(&self, size: usize) -> Amount {
+        let size = u128::try_from(size).expect("compute_fee conversion");
+        Amount::from_atoms(self.atoms_per_kb * size / 1000)
+    }
+
+    pub(crate) fn tokens_per_kb(&self) -> u128 {
+        self.atoms_per_kb
+    }
+
+    fn div_up(fee: u128, tx_size: usize) -> u128 {
+        let tx_size = u128::try_from(tx_size).expect("div_up conversion");
+        (fee + tx_size - 1) / tx_size
```

Comment:
A few things:
- I think the implementation is incorrect. From the call (and from comparing with bitcoin), I see you want to do a `nFeePaid * 1000 / nSize`. For that, you just have to be sure that nSize is non-zero. Assuming you're the caller, maybe that's not a concern. I'm sorry, I'm really lost in the many iterations were done to solve this.
- If you're rounding down, you don't need `-1`, and this can be called div_and_floor, maybe? Not sure. Let's chat about it.

Reply:
This seems like a correct implementation of division with rounding up. It has however some issues with overflow. If `fee` is close to `u128::MAX`, then `fee + tx_size` might overflow. I am not sure if this can happen now but I would still be more comfortable if this function behave correctly around extreme values.

How about something like this?
```rust
let result = fee / tx_size;
let round_up = (fee % tx_size > 0) as u128;
fee + round_up
```

Btw, can we also make the divisor `NonZeroUSize` here?

Reply:
The reason I got confused is that in bitcoin it rounds down. We'll have a discussion about all these operations on Monday.